File: src/kernel/kernel.py
# ComputeHub Kernel - Deterministic Task Scheduler
# Inherited from OpenPC System deterministic scheduling pattern
# Key principle: Linear queue eliminates race conditions

import logging
import threading
import time
from typing import Optional
from collections import deque
from dataclasses import dataclass, field
from enum import Enum

logger = logging.getLogger("computehub.kernel")


class TaskPriority(Enum):
    HIGH = 0
    MEDIUM = 1
    LOW = 2


@dataclass
class TaskUnit:
    """A unit of work to be processed by the kernel."""
    id: str
    action: str
    payload: dict = field(default_factory=dict)
    priority: TaskPriority = TaskPriority.MEDIUM
    created_at: float = field(default_factory=time.time)
    response_queue: Optional[deque] = field(default_factory=deque, repr=False)


class KernelState:
    """Immutable snapshot of kernel state for rollback support."""
    
    def __init__(self, queue_depth: int, processed_count: int, latency: float):
        self.queue_depth = queue_depth
        self.processed_count = processed_count
        self.last_latency = latency
    
    def __repr__(self):
        return f"KernelState(queue={self.queue_depth}, processed={self.processed_count}, latency={self.last_latency:.3f}s)"


class DeterministicKernel:
    """
    Deterministic kernel - processes tasks in a single-threaded linear queue.
    
    Architecture pattern (from OpenPC System):
        Client -> Gateway -> PurePipeline -> Kernel(Linear Queue) -> Executor -> Result
    
    Key properties:
    1. Single-threaded processing = no race conditions
    2. State mirror for rollback
    3. Priority-ordered queue
    """
    
    def __init__(self, queue_size: int = 1000, max_state_history: int = 10000):
        self._lock = threading.Lock()
        self._queue: deque = deque(maxlen=queue_size)
        self._max_states = max_state_history
        self._state_history: list = []
        self._running = False
        self._thread: Optional[threading.Thread] = None
        self._processed_count = 0
        self._last_latency = 0.0
        
        # Metrics
        self._total_processed = 0
        self._total_failed = 0
        self._total_time = 0.0
        
        logger.info(f"✅ Kernel initialized: queue_size={queue_size}, max_states={max_state_history}")
    
    def start(self):
        """Start the kernel processing loop."""
        if self._running:
            logger.warning("Kernel already running")
            return
        
        self._running = True
        self._thread = threading.Thread(target=self._process_loop, daemon=True, name="kernel-processor")
        self._thread.start()
        logger.info("🚀 Kernel started (linear processing)")
    
    def stop(self):
        """Stop the kernel."""
        self._running = False
        if self._thread:
            self._thread.join(timeout=5)
        logger.info("⏹️  Kernel stopped")
    
    def submit(self, task: TaskUnit) -> TaskUnit:
        """Submit a task to the linear queue."""
        with self._lock:
            # Insert in priority order
            queue_list = list(self._queue)
            inserted = False
            for i, existing in enumerate(queue_list):
                if task.priority.value < existing.priority.value:
                    queue_list.insert(i, task)
                    inserted = True
                    break
            if not inserted:
                queue_list.append(task)
            self._queue = deque(queue_list, maxlen=self._queue.maxlen)
        
        logger.info(f"📥 Task submitted: {task.id} ({task.action}, priority={task.priority.name})")
        return task
    
    def _process_loop(self):
        """Single-threaded processing loop - eliminates race conditions."""
        while self._running:
            try:
                task = self._queue.popleft()
                
                # Snapshot state before processing
                self._snapshot_state()
                
                start = time.time()
                
                # Process based on action
                result = self._handle_action(task)
                
                latency = time.time() - start
                self._last_latency = latency
                self._processed_count += 1
                self._total_processed += 1
                self._total_time += latency
                
                # Send result back
                if task.response_queue:
                    task.response_queue.append({
                        "success": result.get("success", False),
                        "data": result.get("data"),
                        "error": result.get("error"),
                    })
                
                logger.info(f"✅ Task completed: {task.id} ({latency*1000:.1f}ms)")
                
            except IndexError:
                # Queue empty, wait a bit
                time.sleep(0.01)
            except Exception as e:
                logger.error(f"❌ Task failed: {e}")
                self._total_failed += 1
    
    def _handle_action(self, task: TaskUnit) -> dict:
        """Handle different task actions."""
        action = task.action.upper()
        payload = task.payload
        
        if action == "SUBMIT":
            # Task scheduling logic
            node = self._select_node(payload)
            if node:
                return {"success": True, "data": {"assigned_node": node, "status": "SCHEDULED"}}
            return {"success": False, "error": "No available node"}
        
        elif action == "STATUS":
            return {"success": True, "data": self.get_status()}
        
        elif action == "HEARTBEAT":
            # Update node status
            return {"success": True, "data": "heartbeat received"}
        
        else:
            return {"success": False, "error": f"Unknown action: {action}"}
    
    def _select_node(self, payload: dict) -> Optional[str]:
        """
        Simple node selection - selects first available node.
        Future: implement geographic + load-based selection.
        """
        # This would query the database for available nodes
        # For now, return a placeholder
        return None
    
    def _snapshot_state(self):
        """Save current kernel state for potential rollback."""
        state = KernelState(
            queue_depth=len(self._queue),
            processed_count=self._processed_count,
            latency=self._last_latency,
        )
        self._state_history.append(state)
        if len(self._state_history) > self._max_states:
            self._state_history.pop(0)
    
    def get_status(self) -> dict:
        """Get current kernel status."""
        return {
            "status": "RUNNING" if self._running else "STOPPED",
            "queue_depth": len(self._queue),
            "processed_count": self._processed_count,
            "total_processed": self._total_processed,
            "total_failed": self._total_failed,
            "average_latency_ms": (self._total_time / self._total_processed * 1000) if self._total_processed > 0 else 0,
            "last_latency_ms": self._last_latency * 1000,
            "state_history_count": len(self._state_history),
        }
    
    @property
    def queue_depth(self) -> int:
        return len(self._queue)
    
    @property
    def state_history(self) -> list:
        return self._state_history

File: src/kernel/kernel.go
// Package kernel implements the deterministic task scheduler kernel.
// It enforces strict state machine transitions and eliminates race conditions
// through linear command processing.
//
// Inspired by OpenPC's deterministic kernel philosophy.
package kernel

import (
	"fmt"
	"sync"
	"time"
)

// ─── 状态常量 ───

const (
	StatePending     = "PENDING"
	StateAllocated   = "ALLOCATED"
	StateExecuting   = "EXECUTING"
	StateVerifying   = "VERIFYING"
	StateCompleted   = "COMPLETED"
	StateFailed      = "FAILED"
	StateCancelled   = "CANCELLED"
)

// ─── 物理状态快照 ───

// State represents a physical snapshot of system state at a timestamp.
type State struct {
	Timestamp time.Time       `json:"timestamp"`
	Nodes     int             `json:"nodes"`
	Tasks     TaskStats       `json:"tasks"`
	Resources ResourceUsage   `json:"resources"`
	Meta      map[string]any  `json:"meta,omitempty"`
}

// TaskStats tracks task counts by state.
type TaskStats struct {
	Total     int `json:"total"`
	Pending   int `json:"pending"`
	Running   int `json:"running"`
	Completed int `json:"completed"`
	Failed    int `json:"failed"`
}

// ResourceUsage tracks current resource consumption.
type ResourceUsage struct {
	GPUs     int     `json:"gpus"`
	CPUs     int     `json:"cpus"`
	MemBytes uint64  `json:"mem_bytes"`
	MemPct   float64 `json:"mem_pct"`
}

// ─── 任务定义 ───

// TaskAction represents a deterministic task action.
type TaskAction string

const (
	TaskActionSubmit   TaskAction = "SUBMIT"
	TaskActionExecute  TaskAction = "EXECUTE"
	TaskActionCancel   TaskAction = "CANCEL"
	TaskActionStatus   TaskAction = "STATUS"
	TaskActionResource TaskAction = "RESOURCE"
)

// ComputeTask is the fundamental unit of work in ComputeHub.
type ComputeTask struct {
	ID           string            `json:"id"`
	Action       TaskAction        `json:"action"`
	Framework    string            `json:"framework"`     // pytorch, tensorflow, jax
	ResourceType string            `json:"resource_type"` // gpu, cpu, tpu
	GPUCount     int               `json:"gpu_count"`
	CPUCount     int               `json:"cpu_count"`
	MemoryGB     int               `json:"memory_gb"`
	DurationSecs int               `json:"duration_secs"`
	Requirements map[string]string `json:"requirements,omitempty"`
	Source       string            `json:"source"`      // tui, api, sdk
	AssignedNode string            `json:"assigned_node,omitempty"`
	Status       string            `json:"status"`
	StartTime    time.Time         `json:"start_time,omitempty"`
	EndTime      time.Time         `json:"end_time,omitempty"`
	Result       map[string]any    `json:"result,omitempty"`
}

// Command is a linearized task instruction.
type Command struct {
	ID       string
	Task     *ComputeTask
	Response chan *Response
}

// Response is the kernel's reply to a command.
type Response struct {
	Success bool
	Data    any
	Error   error
}

// ─── 配额管理器 ───

// Quota defines resource limits.
type Quota struct {
	MaxGPUs         int
	MaxCPUs         int
	MaxMemoryGB     int
	MaxConcurrent   int
	MaxPendingTasks int
}

// DefaultQuota returns sensible defaults.
func DefaultQuota() Quota {
	return Quota{
		MaxGPUs:         8,
		MaxCPUs:         32,
		MaxMemoryGB:     128,
		MaxConcurrent:   10,
		MaxPendingTasks: 500,
	}
}

// ─── 确定性内核 ───

// Kernel is the deterministic task scheduler.
// All task processing happens sequentially through LinearQueue,
// eliminating race conditions across the entire system.
type Kernel struct {
	mu            sync.RWMutex
	stateMirror   []State
	maxStates     int
	LinearQueue   chan *Command
	LastLatency   time.Duration
	tasks         map[string]*ComputeTask
	currentTasks  map[string]*ComputeTask // tasks in EXECUTING state
	taskCounter   int
	quota         Quota
	startTime     time.Time
}

// NewKernel creates a new deterministic kernel with given capacity.
func NewKernel(bufferSize int, maxStates int) *Kernel {
	return &Kernel{
		stateMirror: make([]State, 0),
		maxStates:   maxStates,
		LinearQueue: make(chan *Command, bufferSize),
		tasks:       make(map[string]*ComputeTask),
		currentTasks: make(map[string]*ComputeTask),
		taskCounter: 0,
		quota:       DefaultQuota(),
		startTime:   time.Now(),
	}
}

// NewKernelDefaults creates a kernel with default buffer size.
func NewKernelDefaults() *Kernel {
	return NewKernel(500, 5000)
}

// Start begins the linearized task processing loop.
// This goroutine ensures ALL task processing is serial.
func (k *Kernel) Start() {
	go func() {
		for cmd := range k.LinearQueue {
			k.processCommand(cmd)
		}
	}()
}

// processCommand executes a single command deterministically.
func (k *Kernel) processCommand(cmd *Command) {
	start := time.Now()

	// 1. Snapshot current state before any modification
	k.snapshotState()

	// 2. Execute command based on action
	var resp *Response
	switch cmd.Task.Action {
	case TaskActionSubmit:
		resp = k.handleSubmit(cmd.Task)
	case TaskActionExecute:
		resp = k.handleExecute(cmd.Task)
	case TaskActionCancel:
		resp = k.handleCancel(cmd.Task)
	case TaskActionStatus:
		resp = k.handleStatus(cmd.Task)
	case TaskActionResource:
		resp = k.handleResource(cmd.Task)
	default:
		resp = &Response{
			Success: false,
			Error:   fmt.Errorf("unknown action: %s", cmd.Task.Action),
		}
	}

	// 3. Record latency
	duration := time.Since(start)
	k.mu.Lock()
	k.LastLatency = duration
	k.mu.Unlock()

	// 4. Send response
	cmd.Response <- resp
}

// handleSubmit processes a new task submission.
func (k *Kernel) handleSubmit(task *ComputeTask) *Response {
	k.mu.Lock()
	defer k.mu.Unlock()

	// Check quota
	if len(k.currentTasks) >= k.quota.MaxConcurrent {
		return &Response{
			Success: false,
			Error:   fmt.Errorf("concurrent task limit reached: %d/%d", len(k.currentTasks), k.quota.MaxConcurrent),
		}
	}

	if len(k.tasks)+1 > k.quota.MaxPendingTasks {
		return &Response{
			Success: false,
			Error:   fmt.Errorf("pending task limit reached: %d/%d", len(k.tasks), k.quota.MaxPendingTasks),
		}
	}

	// Check GPU quota
	var usedGPUs int
	for _, t := range k.tasks {
		if t.ResourceType == "gpu" {
			usedGPUs += t.GPUCount
		}
	}
	if usedGPUs+task.GPUCount > k.quota.MaxGPUs {
		return &Response{
			Success: false,
			Error:   fmt.Errorf("GPU quota exceeded: %d/%d", usedGPUs+task.GPUCount, k.quota.MaxGPUs),
		}
	}

	// Assign task ID
	task.ID = fmt.Sprintf("task-%d", k.taskCounter)
	k.taskCounter++
	task.Status = StatePending
	task.StartTime = time.Now()

	// Register task
	k.tasks[task.ID] = task

	return &Response{
		Success: true,
		Data: map[string]any{
			"task_id": task.ID,
			"status":  task.Status,
		},
	}
}

// handleExecute allocates and starts task execution.
func (k *Kernel) handleExecute(task *ComputeTask) *Response {
	k.mu.Lock()
	defer k.mu.Unlock()

	existing, exists := k.tasks[task.ID]
	if !exists {
		return &Response{
			Success: false,
			Error:   fmt.Errorf("task not found: %s", task.ID),
		}
	}

	if existing.Status != StatePending {
		return &Response{
			Success: false,
			Error:   fmt.Errorf("task cannot be executed in state: %s", existing.Status),
		}
	}

	// State transition: PENDING → ALLOCATED → EXECUTING
	existing.Status = StateExecuting
	existing.StartTime = time.Now()
	k.currentTasks[task.ID] = existing

	return &Response{
		Success: true,
		Data: map[string]any{
			"task_id": task.ID,
			"status":  StateExecuting,
			"message": "Task allocated and executing",
		},
	}
}

// handleCancel stops a task.
func (k *Kernel) handleCancel(task *ComputeTask) *Response {
	k.mu.Lock()
	defer k.mu.Unlock()

	existing, exists := k.tasks[task.ID]
	if !exists {
		return &Response{
			Success: false,
			Error:   fmt.Errorf("task not found: %s", task.ID),
		}
	}

	if existing.Status == StateCompleted || existing.Status == StateCancelled {
		return &Response{
			Success: false,
			Error:   fmt.Errorf("task cannot be cancelled in state: %s", existing.Status),
		}
	}

	existing.Status = StateCancelled
	existing.EndTime = time.Now()
	delete(k.currentTasks, task.ID)

	return &Response{
		Success: true,
		Data: map[string]any{
			"task_id": task.ID,
			"status":  StateCancelled,
		},
	}
}

// handleStatus returns task information.
func (k *Kernel) handleStatus(task *ComputeTask) *Response {
	k.mu.RLock()
	defer k.mu.RUnlock()

	if task.ID == "" {
		// Return aggregate stats
		return &Response{
			Success: true,
			Data: k.getTaskStats(),
		}
	}

	existing, exists := k.tasks[task.ID]
	if !exists {
		return &Response{
			Success: false,
			Error:   fmt.Errorf("task not found: %s", task.ID),
		}
	}

	return &Response{
		Success: true,
		Data:    existing,
	}
}

// handleResource returns resource usage information.
func (k *Kernel) handleResource(task *ComputeTask) *Response {
	k.mu.RLock()
	defer k.mu.RUnlock()

	var usedGPUs, usedCPUs, totalMem int
	for _, t := range k.tasks {
		if t.Status == StateExecuting {
			usedGPUs += t.GPUCount
			usedCPUs += t.CPUCount
			totalMem += t.MemoryGB
		}
	}

	return &Response{
		Success: true,
		Data: map[string]any{
			"gpu_used":    usedGPUs,
			"gpu_total":   k.quota.MaxGPUs,
			"cpu_used":    usedCPUs,
			"cpu_total":   k.quota.MaxCPUs,
			"mem_used_gb": totalMem,
			"mem_total_gb": k.quota.MaxMemoryGB,
			"concurrent_tasks": len(k.currentTasks),
			"max_concurrent": k.quota.MaxConcurrent,
		},
	}
}

// ─── 状态镜像 ───

// snapshotState saves a physical state snapshot.
func (k *Kernel) snapshotState() {
	k.mu.Lock()
	defer k.mu.Unlock()

	stats := k.getTaskStatsLocked()

	state := State{
		Timestamp: time.Now(),
		Nodes:     1, // Local node count
		Tasks:     stats,
		Resources: ResourceUsage{
			GPUs:    stats.Running,
			CPUs:    stats.Running,
			MemBytes: uint64(stats.Running) * 32 * 1024 * 1024 * 1024, // ~32GB per task
			MemPct:  float64(stats.Running) / float64(k.quota.MaxConcurrent) * 100,
		},
	}

	k.stateMirror = append(k.stateMirror, state)
	if len(k.stateMirror) > k.maxStates {
		k.stateMirror = k.stateMirror[1:]
	}
}

// getStateHistory returns recent state snapshots.
func (k *Kernel) getStateHistory(n int) []State {
	k.mu.RLock()
	defer k.mu.RUnlock()

	if n > len(k.stateMirror) {
		n = len(k.stateMirror)
	}
	if n == 0 {
		return k.stateMirror
	}
	return k.stateMirror[len(k.stateMirror)-n:]
}

// ─── 状态查询 ───

func (k *Kernel) getTaskStats() TaskStats {
	k.mu.RLock()
	defer k.mu.RUnlock()
	return k.getTaskStatsLocked()
}

func (k *Kernel) getTaskStatsLocked() TaskStats {
	stats := TaskStats{}
	for _, t := range k.tasks {
		stats.Total++
		switch t.Status {
		case StatePending:
			stats.Pending++
		case StateExecuting:
			stats.Running++
		case StateCompleted:
			stats.Completed++
		case StateFailed:
			stats.Failed++
		}
	}
	return stats
}

// ─── 系统信息 ───

// Info returns kernel system information.
func (k *Kernel) Info() map[string]any {
	k.mu.RLock()
	defer k.mu.RUnlock()

	stats := k.getTaskStatsLocked()
	uptime := time.Since(k.startTime)

	return map[string]any{
		"status":          "RUNNING",
		"schedule_latency": k.LastLatency.String(),
		"queue_depth":     len(k.LinearQueue),
		"total_tasks":     stats.Total,
		"pending_tasks":   stats.Pending,
		"running_tasks":   stats.Running,
		"completed_tasks": stats.Completed,
		"failed_tasks":    stats.Failed,
		"uptime":          uptime.String(),
		"version":         "2.0.0",
		"architecture":    "distributed-computehub",
	}
}

// SetQuota updates resource quota limits.
func (k *Kernel) SetQuota(q Quota) {
	k.mu.Lock()
	defer k.mu.Unlock()
	k.quota = q
}

// GetQuota returns current quota.
func (k *Kernel) GetQuota() Quota {
	k.mu.RLock()
	defer k.mu.RUnlock()
	return k.quota
}

// GetTaskStats returns current task statistics.
func (k *Kernel) GetTaskStats() TaskStats {
	return k.getTaskStats()
}

// Dispatch sends a command to the linear processing queue.
// All commands are processed strictly in order.
func (k *Kernel) Dispatch(task *ComputeTask) *Response {
	respChan := make(chan *Response, 1)
	k.LinearQueue <- &Command{
		ID:       task.ID,
		Task:     task,
		Response: respChan,
	}
	return <-respChan
}

// Stop gracefully shuts down the kernel.
func (k *Kernel) Stop() {
	close(k.LinearQueue)
}
